Add database

rovniipoc opened the pull request:
Добрый день.
Представляю на ревью решение по ТЗ 12го спринта.

rovniipoc committed:
Добавить BaseRepository и UserDbStorage

rovniipoc committed:
Добавить новые сущности и DAL

rovniipoc committed:
Черновой вариант решения ТЗ

rovniipoc committed:
Добавить немного тестов

rovniipoc committed:
Поправить кодстайл

rovniipoc committed:
Доработать тесты

rovniipoc committed:
Добавить application-test.properties

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -57,6 +54,11 @@ public Film delete(@Valid @RequestBody Film film) {
 
+    @DeleteMapping("/all")
+    public void deleteAll() {
+        filmService.deleteAll();
```
логирвоание

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -43,10 +43,7 @@ public Collection<Film> getPopularFilms(@RequestParam(defaultValue = "10") Long
-        Film response = filmService.create(film);
-        log.info("Отправлен ответ Post /films с телом {}", response);
-        return response;
+        return filmService.create(film);
```
Почему убрал логирвоание?

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/GenreController.java:
```diff
@@ -0,0 +1,31 @@
+
+    @GetMapping("/{id}")
+    public Genre findById(@PathVariable Long id) {
+        return  genreService.getGenreById(id);
```
лишний пробел

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/GenreController.java:
```diff
@@ -0,0 +1,31 @@
+import ru.yandex.practicum.filmorate.service.GenreService;
+
+import java.util.Collection;
+
```
Нет логирвоания

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/GenreController.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    @DeleteMapping("/all")
+    public void deleteAll() {
```
В задании про этот метод ничего не сказано, не стоит его делать, тем более что добавить их нельзя, если кто то удалит то все жанры просто исчезнут

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/MpaController.java:
```diff
@@ -0,0 +1,31 @@
+import ru.yandex.practicum.filmorate.service.MpaService;
+
+import java.util.Collection;
+
```
Логирование

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/MpaController.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    @DeleteMapping("/all")
+    public void deleteAll() {
```
Этот метод не нужен

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -82,11 +82,8 @@ public User update(@Valid @RequestBody User user) {
-        log.info("Отправлен ответ Put /users/{}/friends/{} с телом: {}", id, friendId, response);
-        return response;
+    public void makeFriend(@PathVariable Long id, @PathVariable Long friendId) {
+        userService.makeFriends(id, friendId);
```
Почему убрал логирование?

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -46,7 +49,11 @@ public Collection<Film> getAll() {
-        return filmStorage.get(id);
+        try {
+            return filmStorage.get(id);
+        } catch (RuntimeException e) {
```
RuntimeException e
и NotFount это совсем разные по смыслу исключения
Тут лучше RuntimeException не отлавливать, все что угодно может быть под этим исключением, а ты будешь возвращать не найден - но это не так

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -12,26 +14,27 @@
+        return filmStorage.get(filmId);
     }
 
     public Film unlike(Long filmId, Long userId) {
```
Можно не возвращать фильм, просто void
Лайк убрали и все

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -12,26 +14,27 @@
+        this.userStorage = userStorage;
+    }
+
     public Film like(Long filmId, Long userId) {
```
Можно не возвращат ь фильм, просто void
Лайк поставили и все

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+    private final LikeDbStorage likeDbStorage;
+    private final MpaDbStorage mpaDbStorage;
+    private final GenreDbStorage genreDbStorage;
+    private final FilmGenreDbStorage filmGenreDbStorage;
```
Storage не вызывать другие стораджи. 
Они только работают с обьектом Film и все. Никаких других зависимостей делать не нужно, если нужно где-то что то вызвать, это вызывается из севриса. 
Поэтому все другие стораджи от сюда стоит убрать
    private final LikeDbStorage likeDbStorage;
    private final MpaDbStorage mpaDbStorage;
    private final GenreDbStorage genreDbStorage;
    private final FilmGenreDbStorage filmGenreDbStorage;

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+                film.getMpa().getId()
+        );
+        if (!film.getGenres().isEmpty()) {
+            filmGenreDbStorage.addManyGenresToFilm(id,
```
Вызывай не мотед другого хранилища, а выполняй запрос для связи жанров и фильмов
Выполнять запросы в цикле это не очень, используй батчи
```
	private void saveGenres(Film film) {
		final Long filmId = film.getId();
		jdbcTemplate.update("delete from FILM_GENRES where FILM_ID = ?", filmId);
		final Set<Genre> genres = film.getGenres();
		if (genres == null || genres.isEmpty()) {
			return;
		}
		final ArrayList<Genre> genreList = new ArrayList<>(genres);
		jdbcTemplate.batchUpdate(
				"insert into FILM_GENRES (FILM_ID, GENRE_ID) values (?, ?)",
				new BatchPreparedStatementSetter() {
					public void setValues(PreparedStatement ps, int i) throws SQLException {
						ps.setLong(1, filmId);
						ps.setLong(2, genreList.get(i).getId());
					}

					public int getBatchSize() {
						return genreList.size();
					}
				});
	}
```

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmGenreDbStorage.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public void addManyGenresToFilm(Long filmId, Collection<Long> genreIds) {
+        for (Long genreId : genreIds) {
```
Выполянть запросы ы цикле это не очень, на каждый запрос создается подключение, можно убрать этот метод позже покажу как стоит сделать

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    @Override
+    public Film remove(Film film) {
```
Для методов удаления лучше возвращать void

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -20,10 +19,8 @@ public class Film {
     @Positive
     private Long duration;
-    @JsonIgnore
     private Set<Long> likes = new HashSet<>();
```
Лучше сделать это поле
	private long rate;
и подсчитывать рейтинг как число а не возвращать список идентификаторов

YazykovMikhail replied:
Оставил комментарий в LikeDbStorage

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/LikeDbStorage.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.Collection;
+
+@Repository
+public class LikeDbStorage extends BaseRepository<Like> {
```
Т.е вот как будет выглядеть бд для работы с лайками:

```
@Component
public class DbLikesStorage implements LikesStorage {

	private final JdbcTemplate jdbcTemplate;

	public DbLikesStorage(JdbcTemplate jdbcTemplate) {
		this.jdbcTemplate = jdbcTemplate;
	}

	@Override
	public void addLike(long filmId, long userId) {
		String sqlQuery = "merge into LIKES (film_id, user_id) values (?, ?)";
		jdbcTemplate.update(sqlQuery, filmId, userId);
		updateRate(filmId);
	}

	private void updateRate(long filmId) {
		String sqlQuery = "update FILMS f set rate = (select count(l.user_id) from LIKES l where l.film_id = f.film_id)  where film_id = ?";
		jdbcTemplate.update(sqlQuery, filmId);
	}

	@Override
	public void removeLike(long filmId, long userId) {
		String sqlQuery = "delete from LIKES where FILM_ID = ? and USER_ID = ?";
		jdbcTemplate.update(sqlQuery, filmId, userId);
		updateRate(filmId);
	}

	@Override
	public List<Film> getPopular(int count) {
		final String sqlQuery = "select * from FILMS f, MPA m where f.MPA_ID = m.MPA_ID  order by RATE desc limit ?";
		return jdbcTemplate.query(sqlQuery, DbFilmStorage::makeFilm, count);
	}

}
```
Обрати внимание на метод updateRate который вызывается для актуализации рейтинга - поля в фильм

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+    public Film get(Long id) {
+        Film film = findOne(FIND_BY_ID_QUERY, id);
+        if (film != null) {
+            film.getLikes().addAll(likeDbStorage.findLikesByFilm(film).stream()
```
Это будет не нужно, просто в запросе нужно будет выводить еще одно поле RATE посмотри комментарии к LikeDbStorage

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+                    .map(FilmGenre::getGenreId)
+                    .map(genreDbStorage::getGenre)
+                    .toList());
+            film.getMpa().setName(mpaDbStorage.getMpa(film.getMpa().getId()).getName());
```
Не нужно для этого делать отедльрный запрос и вызов, просто при получении фильма, получай сразу и mpa 

`		final String sqlQuery = "select * from FILMS f, MPA m where f.MPA_ID = m.MPA_ID AND FILM_ID = ?";
`

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,133 @@
+        if (film != null) {
+            film.getLikes().addAll(likeDbStorage.findLikesByFilm(film).stream()
+                    .map(Like::getUserId).toList());
+            film.getGenres().addAll(filmGenreDbStorage.findGenresByFilm(film).stream()
```
Как ты уже понял, делать запрос в цикле это не очень, и вызывать один репозиторий из другого тоже.
Поэтому в сервисе покажу как получить данные, когда идет связь многие ко многим

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -46,7 +49,11 @@ public Collection<Film> getAll() {
     }
 
     public Film getFilmById(Long id) {
```
public Film get(long id) {
		final Film film = super.get(id);
		genreStorage.load(Collections.singletonList(film));
		return film;
	}
Collections.singletonList(film) - сделано для обратной совместимочти что бы этот же метод можно было использовать когда нужно заполнить данные не по одному фильму, а по списку, когда мы вызываем getFilms

YazykovMikhail replied:
А вот так будет выглядть сам метод load 
```
	public void load(List<Film> films) {
		String inSql = String.join(",", Collections.nCopies(films.size(), "?"));
		final Map<Long, Film> filmById = films.stream().collect(Collectors.toMap(Film::getId, identity()));
		final String sqlQuery = "select * from GENRES g, FILM_GENRES fg where fg.GENRE_ID = g.GENRE_ID AND fg.FILM_ID in (" + inSql + ")";
		jdbcTemplate.query(sqlQuery, (rs) -> {
			final Film film = filmById.get(rs.getLong("FILM_ID"));
			film.addGenre(makeGenre(rs, 0));
		}, films.stream().map(Film::getId).toArray());
	}
```
который находится в GenreDbStorage

YazykovMikhail replied:
Соответсвенно поправь и методы getAll

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/GenreDbStorage.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.Collection;
+
+@Repository
+public class GenreDbStorage extends BaseRepository<Genre> {
```
Тут лолжны быть методы только 
get
getAll
и 
load - котрый будет наполнять фильмы

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/MpaDbStorage.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.Collection;
+
+@Repository
+public class MpaDbStorage extends BaseRepository<Mpa> {
```
Как скзаано в задании 
Тут лолжны быть методы только
get
getAll

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/UserDbStorage.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    @Override
+    public void addFriend(Long userId, Long friendId) {
```
Эти методы тут не нужны, раз есть  friendshipDbStorage
кому нужно пусть вызвают от туда.

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -19,6 +18,5 @@ public class User {
     @Past
     private LocalDate birthday;
-    @JsonIgnore
     private Set<Long> friends = new HashSet<>();
```
Теперь это поле тут не нужно, раньше оно было утт для хранения, а теперь эти данные храняться в бд

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/UserDbStorage.java:
```diff
@@ -0,0 +1,127 @@
+    public User get(Long id) {
+        User user = findOne(FIND_BY_ID_QUERY, id);
+        if (user != null) {
+            user.getFriends().addAll(friendshipDbStorage.findAllFriendsIdByUser(user.getId())
```
Этого не нужно, этого поля не будет в User раньше оно там было, что бы можно было удобно хранить, теперь для этого есть таблица

rovniipoc committed:
Добавить логирование в контроллерах, удалить лишние методы в контроллерах

rovniipoc committed:
Изменить возвращаемый тип на void в методах like(), unlike() фильм-сервиса (и контроллера)

rovniipoc committed:
Исправлен метод add(Film film) FilmDbStorage

 Добавлен метод saveGenres(Film film) для загрузки полученных данных о жанрах из film в sql таблицу film_genres

rovniipoc committed:
Изменить FilmDbStorage, LikeDbStorage

Исправлен метод update(Film film) FilmDbStorage (через saveGenres)
Перенесены методы like из FilmDbStorage в LikeDbStorage (теперь FilmService в части добавления или удаления лайков обращается к LikeDbStorage)

rovniipoc committed:
Изменить метод get() FilmDbStorage и прочие изменения

В методе get():
-убран стрим для получения лайков, заменен на получение rate (кол-во лайков) через sql-запрос;
-убран стрим для получения mpa, заменен на получение id и mpa_name через sql-запрос, объект Mpa для Film формируется на основе полученных данных через FilmRowMapper

Сопутствующие изменения:
-в схеме в таблице ratings (mpa) изменено поле с name на mpa_name;
-FilmRowMapper теперь использует в т.ч. mpa_name, полученный через sql-запрос
-FilmRowMapper теперь использует в т.ч. rate (кол-во лайков), полученный через sql-запрос
-в схеме в таблице genres изменено поле с name на genres_name;

rovniipoc committed:
Изменить метод getFilmById() FilmService, изменить GenreDbStorage

В GenreDbStorage:
-добавлен метод load;
-убраны все методы кроме get, getAll, load;

rovniipoc committed:
Изменить FilmDbStorage

В FilmDbStorage:
-убран метод removeGenresFromFilm();
-убран методы addGenreToFilm() и addManyGenresToFilm(), этот функционал теперь выполняется в GenreDbStorage метод load();

rovniipoc committed:
Изменить FilmDbStorage

В FilmDbStorage:
-из метода get() убрана подгрузка жанров, теперь этот функционал в методе getFilmById() FilmService;

rovniipoc committed:
Изменить FilmService

В FilmService:
-в методе getPopularFilms() теперь используется метод getPopular() из LikeDbStorage

rovniipoc committed:
Изменить MpaDbStorage

В MpaDbStorage:
-удалены все методы кроме getMpa() и findAllMpas()

rovniipoc committed:
Изменить UserService

В UserService:
-в методах makeFriend() и unfriend() теперь используется FriendshipDbStorage

rovniipoc committed:
Изменить User

В User:
-удалено поле friends;

rovniipoc committed:
Изменить UserDbStorage, UserService, UserStorage, GenreService

В UserDbStorage:
-метод get() возвращает user без наполнения друзьями, выдача друзей теперь происходит через getUserFriends() UserDbStorage;
-удалены методы addFriend(), removeFriend(), этот функционал в FriendshipDbStorage;

В UserService:
-исправлен (оптимизирован) метод getFriends();

В UserStorage:
-удалены интерфейсные методы добавления и удаления друзей, т.к. теперь это функционал FriendshipDbStorage;

В GenreService:
-удален метод ссылающийся на удаленный ранее метод удаления всех жанров класс GenreDbStorage;

rovniipoc committed:
Реализовать требуемый функционал
